### ISSUE TITLE
[Core] Slow Embedded Object Ref Counting

### ISSUE THREAD

**Issue by pdames** [ray-project/ray#17803]
### What is the problem?
`ray.get()` shows unexpectedly high latencies when getting objects refs that contain embedded object refs. I've run the repro script below to measure the object ref counting latency here, which produced the following `ray.get()` latency results:
```
1000 object refs with 600 embedded refs: ~10 seconds w/o ref hiding, ~0.5 seconds w/ ref hiding
2000 object refs with 600 embedded refs: ~20 seconds w/o ref hiding, ~1 seconds w/ ref hiding
2000 object refs with 1200 embedded refs: ~40 seconds w/o ref hiding, ~1 seconds w/ ref hiding
500 object refs with 600 embedded refs: ~5 seconds w/o ref hiding, ~0.5 seconds w/ ref hiding
500 object refs with 1200 embedded refs: ~10 seconds w/o ref hiding, ~0.5 seconds w/ ref hiding
500 object refs with 2400 embedded refs: ~20 seconds w/o ref hiding, ~0.5 seconds w/ ref hiding
```

Here, "ref hiding" refers to using `cloudpickle.dumps(obj_ref)` to hide embedded object references from Ray's reference counter. The above figures also show that the runtime is approximately `O(obj_refs*embedded_refs_per_obj)` when embedded object references are not hidden, and roughly `O(obj_refs)` (with an apparent lower-bound of ~0.5 seconds)   when hiding object references. 

*Ray version and other system information (Python version, TensorFlow version, OS):* 
Ray 1.5.1 on AWS Deep Learning AMI (Ubuntu 18.04) V30.0

### Reproduction (REQUIRED)
You can configure the `OBJ_REFS`, `EMBEDDED_OBJ_REFS`, and `REF_HIDING` variables of the below script to reproduce the above results, and uncomment the profiler code to trace the issue back to its root cause.

Python Script:
```python
import datetime
import time
import ray
from ray import ray_constants
from ray import cloudpickle


OBJ_REFS = 1000  # number of put_object tasks to run, and number of top-level object refs input to ray.get()
EMBEDDED_OBJ_REFS = 600  # number of embedded object refs per input object
REF_HIDING = False  # whether to hide embedded object refs via cloudpickle.dumps(ray.put(input_object)))


def run():
    cluster_resources = ray.cluster_resources()
    next_node_idx = 0
    node_idx_to_id = {}
    for resource_name in cluster_resources.keys():
        if resource_name.startswith("node:"):
            node_idx_to_id[next_node_idx] = resource_name
            next_node_idx += 1

    input_objects = list(range(OBJ_REFS))
    tasks_pending = []
    i = 0
    for input_object in input_objects:
        node_id = node_idx_to_id[i % len(node_idx_to_id)]
        resources = {node_id: ray_constants.MIN_RESOURCE_GRANULARITY}
        promise = put_object.options(resources=resources).remote(
            input_object,
        )
        tasks_pending.append(promise)
        i += 1
    print(f"getting {len(tasks_pending)} task results...")
    results = ray.get(tasks_pending)
# uncomment for profiling
#   prof = cProfile.Profile()
#   prof = prof.runctx(
#       "ray.get(tasks_pending)",
#       {"ray": ray},
#       {"tasks_pending": tasks_pending}
#   )
#   prof.dump_stats("/home/ubuntu/output.pstats")
    print(f"got {len(results)} task results")
    print(f"retrieved all results at: {datetime.datetime.now()}")


@ray.remote
def put_object(input_object):
    start = time.perf_counter()
    obj_idx_to_obj_id = {}
    for obj_idx in range(EMBEDDED_OBJ_REFS):
        obj_ref = ray.put(
            input_object
        )
        obj_idx_to_obj_id[obj_idx] = obj_ref \
            if not REF_HIDING \
            else cloudpickle.dumps(obj_ref)
    stop = time.perf_counter()
    print("put object latency: ", stop - start)
    print(f"put object end time: {datetime.datetime.now()}")
    return obj_idx_to_obj_id


if __name__ == '__main__':
    print(f"start time: {datetime.datetime.now()}")
    ray.init(address="auto")
    start_e2e = time.perf_counter()
    run()
    stop_e2e = time.perf_counter()
    print("total latency: ", stop_e2e - start_e2e)
    print(f"end time: {datetime.datetime.now()}")
```

Autoscaler Config YAML:
```
cluster_name: ref_counter_repro

max_workers: 1

provider:
    type: aws
    region: us-east-1
    availability_zone: us-east-1a, us-east-1b, us-east-1c, us-east-1d, us-east-1f

auth:
    ssh_user: ubuntu

available_node_types:
  ray.head.default:
    max_workers: 0
    resources: {}
    node_config:
      InstanceType: r5n.8xlarge
      ImageId: latest_dlami

  ray.worker.default:
    min_workers: 1
    max_workers: 1
    resources: {}
    node_config:
      InstanceType: r5n.8xlarge
      ImageId: latest_dlami

head_node_type: ray.head.default

head_start_ray_commands:
    - ray stop
    # r5n-8xlarge worker nodes provide 256GiB of memory spread across 32 CPUs
    # (i.e. an average of 256GiB/32CPU = 8GiB/CPU)
    # The head node reserves 96GiB for the object store and leave 160GiB for
    # everything else (e.g. gcs, driver, tasks, other OS processes, etc.).
    - "ulimit -n 65536; ray start --head --port=6379 --object-manager-port=8076 --object-store-memory=103079215104 --autoscaling-config=~/ray_bootstrap_config.yaml"

worker_start_ray_commands:
    - ray stop
    # r5n-8xlarge worker nodes provide 256GiB of memory spread across 32 CPUs
    # (i.e. an average of 256GiB/32CPU = 8GiB/CPU)
    # Worker nodes reserve 96GiB for the object store and leave 160GiB for
    # everything else (e.g. raylet, tasks, other OS proceses, etc.)
    # 32 CPUs on r5n-8xlarge instances means that we receive an average of
    # 96GiB/32CPU = 3GiB/CPU of object store memory per CPU and
    # 160GiB/32 = 5GiB/CPU of memory for everything else.
    - "ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379 --object-store-memory=103079215104 --object-manager-port=8076"

setup_commands:
  - pip install ray
```
If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [X] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by ericl** [comment]
cc @mwtian

**Comment by mwtian** [comment]
Thanks for the detailed reproduction. I was able to reproduce the behavior with a [modified script](https://gist.github.com/mwtian/625b9a8bdc907e07e175e4d40c21a4a5) to wait for all tasks to finish. The behavior of slow getting and deallocating objects with embedded refs is reproducible on a single local node too.

In the last example, `500 * 2400 = 1.2e6` embedded object refs are created in ~ `20s`. Each `ObjectRef` creation took ~ `17 microsec`. When an embedded `ObjectRef` is created, reference count for the object needs to be incremented. This is where the script spent majority of the walltime. I think there are opportunities to reduce the time spent here. And I will check with others to see if anything can be optimized from a higher level.

Btw, @pdames how many embedded objects refs are you using inside an outer object or in total, and what would be an acceptable latency for calling `ray.get()` on them?

**Comment by pdames** [comment]
@mwtian I'm currently using an average of ~1000 embedded object references per outer object, and a max of ~10,000 embedded object references. In total across all outer objects, I'm producing an average of ~2,000,000 embedded object references, and a max of ~30,000,000 embedded object references. 

An acceptable latency for a common scenario of 625 outer object references with 3520 embedded object references each would be <5 seconds, and ideally <2 seconds.

For a relatively common large-scale scenario of 10,000 outer object references with 8,000 embedded object references each, an acceptable latency would be <20 seconds, and ideally <10 seconds.

**Comment by ericl** [comment]
@mwtian any thoughts on low level optimizations here, or any high level approaches to eliminate the overhead altogether for these bulk processing of references?

**Comment by mwtian** [comment]
@pdames thanks for the info. @ericl one optimization I found is to cache Python stack for all embedded references (#17882). It reduces the processing time by ~50% for objects containing 2400 object references, and by ~65% for objects containing 10000 object references. We need to reduce the time by another ~50% for the common scenario of 625 objects with 3520 embedded object references each, to reach acceptable time here.

For the large-scale scenario of deserializing 10,000 objects containing 8,000 object references under ~20s, we only have ~250 ns to process each embedded object reference, which does not leave room for ref counting logic. To meet the objective, I wonder if we can skip ref counting for the contained object references altogether during deserialization. We may need to have the embedded objects pinned by their outer objects, among additional details. I will check with others for the feasibility. 

@pdames one additional question: are you using the manual "ref hiding" as a workaround right now?

**Comment by ericl** [comment]
Good find on the stack optimization. I wonder where the rest of the time is going, did you get a profile?

250ns is quite a tight budget but not unreasonable--- that's room for more than enough hash table lookups/updates etc. But maybe if we're constructing a bunch of protos etc. it would be too expensive.

**Comment by mwtian** [comment]
Here is a [profile](https://gist.github.com/mwtian/c73d7918c4d4a82429ad2c3939a96d27) for creating and getting 625 objects containing 3520 object refs each, running this [script](https://gist.github.com/mwtian/625b9a8bdc907e07e175e4d40c21a4a5) against master branch. Btw https://www.speedscope.app/ seems to provide a better visualization of `profile_data_collapsed`.

Most of the time spent during deserialization are under `ray::core::CoreWorker::AddLocalReference` and `deserialize_and_register_object_ref`. It seems substantial logic needs to be removed or moved off synchronous path to reduce the time to < 250ns.

**Comment by stephanie-wang** [comment]
@mwtian and I synced offline and I think some steps to try are:
1. Disable recording the call stack entirely.
2. Batching the calls to `deserialize_and_register_object_ref`.

Also, I think the script is not an apples-to-apples comparison because the version that hides the refs doesn't deserialize the ObjectRefs fully. If you deserialize the refs after `ray.get`, it actually has about the same latency as normally embedded ObjectRefs (note, on my laptop). Here's the modified script that I used:
```python
import datetime
import time
import ray
from ray import ray_constants
from ray import cloudpickle


OBJ_REFS = 100  # number of put_object tasks to run, and number of top-level object refs input to ray.get()
EMBEDDED_OBJ_REFS = 600  # number of embedded object refs per input object
REF_HIDING = True # whether to hide embedded object refs via cloudpickle.dumps(ray.put(input_object)))
DESERIALIZE = True # whether to deserialize the hidden embedded object refs


def run():
    cluster_resources = ray.cluster_resources()
    next_node_idx = 0
    node_idx_to_id = {}
    for resource_name in cluster_resources.keys():
        if resource_name.startswith("node:"):
            node_idx_to_id[next_node_idx] = resource_name
            next_node_idx += 1

    input_objects = list(range(OBJ_REFS))
    tasks_pending = []
    i = 0
    for input_object in input_objects:
        node_id = node_idx_to_id[i % len(node_idx_to_id)]
        resources = {node_id: ray_constants.MIN_RESOURCE_GRANULARITY}
        promise = put_object.options(resources=resources).remote(
            input_object,
        )
        tasks_pending.append(promise)
        i += 1
    print(f"getting {len(tasks_pending)} task results...")
    results = ray.get(tasks_pending)
# uncomment for profiling
#   prof = cProfile.Profile()
#   prof = prof.runctx(
#       "ray.get(tasks_pending)",
#       {"ray": ray},
#       {"tasks_pending": tasks_pending}
#   )
#   prof.dump_stats("/home/ubuntu/output.pstats")
    print(f"got {len(results)} task results")
    print(f"retrieved all results at: {datetime.datetime.now()}")

    del results
    start = time.time()
    results = ray.get(tasks_pending)
    if REF_HIDING and DESERIALIZE:
        results = [cloudpickle.loads(result) for result in results]
    end = time.time()
    print("ray.get in", end - start)


@ray.remote
def put_object(input_object):
    start = time.perf_counter()
    obj_idx_to_obj_id = {}
    for obj_idx in range(EMBEDDED_OBJ_REFS):
        obj_ref = ray.put(
            input_object
        )
        obj_idx_to_obj_id[obj_idx] = obj_ref
        #obj_idx_to_obj_id[obj_idx] = obj_ref \
        #    if not REF_HIDING \
        #    else cloudpickle.dumps(obj_ref)
    if REF_HIDING:
        obj_idx_to_obj_id = cloudpickle.dumps(obj_idx_to_obj_id)
    stop = time.perf_counter()
    print("put object latency: ", stop - start)
    print(f"put object end time: {datetime.datetime.now()}")
    return obj_idx_to_obj_id


if __name__ == '__main__':
    print(f"start time: {datetime.datetime.now()}")
    ray.init()
    start_e2e = time.perf_counter()
    run()
    stop_e2e = time.perf_counter()
    print("total latency: ", stop_e2e - start_e2e)
    print(f"end time: {datetime.datetime.now()}")
```

We could definitely get this run time down through the optimizations @mwtian is looking into, but just want to make sure we understand the actual performance gap.

**Comment by pdames** [comment]
@ericl  I'm unsure whether I could improve latency using two layers of serialized objects, since I still need to gather all object refs mapped to a given group ID key across all tasks, then send those object refs to a reduce task for that group ID. Here's a script that more closely reflects the actual scenario:

```python
import datetime
import random
import time
import ray
from ray import ray_constants
from ray import cloudpickle
from collections import defaultdict

ray.init(address="auto")

OBJ_REFS = 1000  # number of put_object tasks to run, and number of top-level object refs input to ray.get()
EMBEDDED_OBJ_REFS = 600  # number of embedded object refs per input object
REF_HIDING = True # whether to hide embedded object refs via cloudpickle.dumps(ray.put(input_object)))


def run():
    cluster_resources = ray.cluster_resources()
    next_node_idx = 0
    node_idx_to_id = {}
    for resource_name in cluster_resources.keys():
        if resource_name.startswith("node:"):
            node_idx_to_id[next_node_idx] = resource_name
            next_node_idx += 1

    group_tasks_pending = []
    i = 0
    start = time.perf_counter()
    for _ in range(OBJ_REFS):
        node_id = node_idx_to_id[i % len(node_idx_to_id)]
        resources = {node_id: ray_constants.MIN_RESOURCE_GRANULARITY}
        promise = put_groups \
            .options(resources=resources) \
            .remote()
        group_tasks_pending.append(promise)
        i += 1
    print(f"getting {len(group_tasks_pending)} groups task results...")
    group_id_to_obj_ref_list = ray.get(group_tasks_pending)
    print(f"got {len(group_id_to_obj_ref_list)} group task results")
    print(f"retrieved all group task results at: {datetime.datetime.now()}")
    stop = time.perf_counter()
    print("total group task time: ", stop - start)

    # gather all object refs for the same task index
    group_id_to_obj_refs = defaultdict(list)
    for group_id_to_obj_ref in group_id_to_obj_ref_list:
        for group_id, obj_ref in group_id_to_obj_ref.items():
            group_id_to_obj_refs[group_id].append(obj_ref)

    reduce_tasks_pending = []
    i = 0
    start = time.perf_counter()
    for group_id, obj_refs in group_id_to_obj_refs.items():
        node_id = node_idx_to_id[i % len(node_idx_to_id)]
        resources = {node_id: ray_constants.MIN_RESOURCE_GRANULARITY}
        promise = reduce_group\
            .options(resources=resources)\
            .remote(group_id, obj_refs)
        reduce_tasks_pending.append(promise)
        i += 1
    print(f"getting {len(reduce_tasks_pending)} reduce tasks...")
    ray.get(reduce_tasks_pending)
    print(f"got all reduce tasks at: {datetime.datetime.now()}")
    stop = time.perf_counter()
    print("total reduce task time: ", stop - start)


@ray.remote
def put_groups():
    start = time.perf_counter()
    pseudorandom_indices = list(range(EMBEDDED_OBJ_REFS))
    random.shuffle(pseudorandom_indices)
    group_id_to_obj_ref = {}
    for group_id in pseudorandom_indices:
        obj_ref = ray.put(group_id)
        group_id_to_obj_ref[group_id] = obj_ref if not REF_HIDING \
            else cloudpickle.dumps(obj_ref)
    stop = time.perf_counter()
    print(f"put {EMBEDDED_OBJ_REFS} objects latency: ", stop - start)
    print(f"put object end time: {datetime.datetime.now()}")
    return group_id_to_obj_ref


@ray.remote
def reduce_group(group_id, obj_refs):
    start = time.perf_counter()
    if REF_HIDING:
        results = [cloudpickle.loads(obj_ref_pkl) for obj_ref_pkl in obj_refs]
        output_objects = ray.get(results)
    else:
        output_objects = ray.get(obj_refs)
    stop = time.perf_counter()
    print(f"get {len(output_objects)} object refs latency: ", stop - start)
    print(f"get object end time: {datetime.datetime.now()}")
    assert(sum(output_objects) == group_id * OBJ_REFS)


if __name__ == '__main__':
    print(f"start time: {datetime.datetime.now()}")
    start_e2e = time.perf_counter()
    run()
    stop_e2e = time.perf_counter()
    print("total latency: ", stop_e2e - start_e2e)
    print(f"end time: {datetime.datetime.now()}")
```

However, I may be misunderstanding how to properly apply this recommendation so, if there's a more optimal way to write the exchange from `put_groups` to `reduce_group` then I'd be happy to try it.

**Comment by mwtian** [comment]
Thanks Patrick for the details. This looks like distributed shuffle of object refs, where minimizing unnecessary object ref deserialization in the driver is needed. Ideally a high-level Ray API can be used for the whole shuffling operation, e.g. Ray Dataset shuffle or ray.experimental.shuffle, so you can save time from getting into implementation details and the underlying Ray implementation would improve with newer versions. Eric and Stephanie probably have better recommendations.

An incremental optimization based on the script [above](https://github.com/ray-project/ray/issues/17803#issuecomment-907726397), is to avoid sending object refs to the driver by using an Actor to run the put operations, store the group id to object ref mappings, return the mappings to the driver, and look up object refs for reduce tasks. e.g. https://gist.github.com/mwtian/140b09ed50fc7f0de172344194b2e7c4.

There can be additional optimizations depending on details e.g.
- The driver can skip the gather and merge step, by having reduce tasks directly ask each map actor for data, if each map actor indeed has almost all groups.
- The map actor can send data instead of object refs directly to the reduce task, if the map actor actually has the data in-memory.

**Comment by ericl** [comment]
I think the two layers of grouping still can help a lot here. Concretely, I would
- in put_groups, return a batched dict of `{group_id: ray.put(refs_for_this_group)}`
- in the driver, gather together all the batched refs for each group, and then,
- pass a list of ref batches to reduce_group instead of a list of refs (you could also launch an intermediate task to combine the batches into a flat list)

Also, is there a reason you need to ray.put() every individual object in put_groups? Batching together multiple objects into ray.put() would reduce the number of references considerably if you have enough memory to buffer them.

**Comment by pdames** [comment]
@ericl By returning a batched dictionary, do you mean returning a separate dictionary for every K of N group IDs, or something else? I'm currently batching together multiple pyarrow tables for each group ID, and then batching together multiple group IDs for a single `ray.put()` in the [production job](https://github.com/ray-project/deltacat/blob/main/deltacat/compute/compactor/compaction_session.py) that is the motivation for this example/repro script, but I'm not returning each batch as a separate dictionary.

To help ensure that the repro scripts don't interfere with mutual understanding [here's an example of an actual compute step utilizing this process](https://github.com/ray-project/deltacat/blob/main/deltacat/compute/compactor/steps/hash_bucket.py#L170-L214). FYI, the delta file envelopes being put are instances of [this model](https://github.com/ray-project/deltacat/blob/main/deltacat/compute/compactor/model/delta_file_envelope.py).

@mwtian With the above context in mind, successful resolution of this issue really winds up looking like comparable latency for an invocation of a [compaction session](https://github.com/ray-project/deltacat/blob/main/deltacat/compute/compactor/compaction_session.py) w/o ref hiding against TiB and PiB-scale datasets.

Right now, ref-hiding is used in the hand-off from the [hash-bucket](https://github.com/ray-project/deltacat/blob/main/deltacat/compute/compactor/steps/hash_bucket.py) step to the [dedupe](https://github.com/ray-project/deltacat/blob/main/deltacat/compute/compactor/steps/dedupe.py) step, and from the dedupe step to the [materialize](https://github.com/ray-project/deltacat/blob/main/deltacat/compute/compactor/steps/materialize.py) step.

I'm also actively pursuing integration of the above project with Ray's Dataset APIs, with the only delays being (1) time to change the implementation to use them and (2) time to run sufficient integration tests to assert that latency, cost, scalability, resilience, and predicted SLA adherence all remain at least comparable to the current implementation. 

To that end, we're also working on exposing an open source implementation of the [DeltaCAT storage APIs ](https://github.com/ray-project/deltacat/blob/main/deltacat/storage/interface.py) so that we can run these compaction sessions in open source and better ensure we haven't _only_ improved a rough repro script w/o improving the real workload (although I think all improvements discussed above will produce demonstrable improvements to a version of the real compaction session workload with ref hiding removed).

**Comment by ericl** [comment]
> @ericl By returning a batched dictionary, do you mean returning a separate dictionary for every K of N group IDs, or something else? I'm currently batching together multiple pyarrow tables for each group ID, and then batching together multiple group IDs for a single ray.put() in the production job that is the motivation for this example/repro script, but I'm not returning each batch as a separate dictionary.

I mean a dictionary of refs. For example, instead of

```
@ray.remote
def gen():
   groups = collections.defaultdict(list)
   for i in range(100000):
       group = i % 100
       ref = ray.put(DUMMY_DATA)
       groups[group].append(ref)
   return groups
   
 ray.get(gen.remote())   # deserializes 100000 references
```

structure it like this

```
@ray.remote
def gen():
   groups = collections.defaultdict(list)
   for i in range(10000):
       group = i % 100
       ref = ray.put(DUMMY_DATA)
       groups[group].append(ref)
   return {k: ray.put(v) for k, v in groups.items()} 
   
 ray.get(gen.remote())   # deserializes 100 references
```

By having two levels, the driver does not need to fully deserialize all the references, just one per group.

**Comment by pdames** [comment]
@ericl Ok, got it - the latter is what I'm doing right now, and my figures for embedded ref counts above refer to the number of groups and not the number of actual elements present in each group (or, to be more specific, my embedded object reference counts above are equal to `total_number_of_groups/X`, since I pack multiple groups together in one `ray.put(group_of_group_ids)` reference to further reduce the number of references sent back to the driver - in my case each reduce task actually handles 1-N roughly-equal-size groups serially rather than just a single group, which was another earlier workaround to help control reference overload on the driver).

**Comment by ericl** [comment]
Got it. In that case then the only workaround I can think of is parallelizing the deserialization by using multiple tasks to do the ref-shuffle instead of just 1.

Right now sounds like the driver is deserializing `|num_mappers| * |num_groups|` refs. But, this could be parallelized with multiple tasks doing the reference grouping (suppose you had K tasks doing the driver work, then each task could handle some fraction of the mapper outputs).

So basically adding a mini "reference-only" shuffle phase to parallelize that bottlenecked portion.

**Comment by pdames** [comment]
Agreed - parallelizing the shuffle instead of bottlenecking it on a single process on the driver sounds promising. I'll start running some experiments to measure the key metrics noted above (latency, cost, scalability, resilience, and predicted SLA adherence) when parallelizing it across K tasks.

**Comment by mwtian** [comment]
Great that we have a higher level strategy to scale the shuffle. Btw we also merged #18078 which reduced the time of deserializing `ObjectRef` by 40~60%. It is available in nightly wheels. I think we can close this issue now.
